tests/org.eclipse.n4js.ui.tests/src/org/eclipse/n4js/tests/externalPackages/GH_986_IntroduceExternalCircularDependencyAndRunTest.java

```diff
@@ -0,0 +1,140 @@
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.n4js.n4mf.ProjectType;
+import org.eclipse.n4js.projectModel.IN4JSCore;
+import org.eclipse.n4js.projectModel.IN4JSProject;
+import org.eclipse.n4js.runner.RuntimeEnvironmentsHelper;
+import org.eclipse.n4js.runner.extension.RuntimeEnvironment;
+import org.eclipse.n4js.tests.builder.AbstractBuilderParticipantTest;
+import org.eclipse.n4js.tests.util.ShippedCodeInitializeTestHelper;
+import org.eclipse.n4js.ui.external.ExternalLibrariesActionsHelper;
+import org.eclipse.n4js.ui.wizard.dependencies.InstallOptions;
+import org.eclipse.n4js.ui.wizard.dependencies.RunnableInstallDependencies;
+import org.eclipse.n4js.utils.StatusHelper;
+import org.eclipse.xtext.ui.testing.util.IResourcesSetupUtil;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import com.google.common.base.Optional;
+import com.google.inject.Inject;
+
+/**
+ * Installs {@code mocha} and {@code eslint} as project dependency, which will introduce a dependency cycle in the
+ * dependency tree of the test project ({@code eslint} and {@code mocha} depend on each other via
+ * {@code devDependencies}).
+ *
+ * This test asserts that despite this dependency cycle, the runner will correctly compute the set of compatible runtime
+ * environment for the test project (for now this is the node.js runner only).
+ */
+public class GH_986_IntroduceExternalCircularDependencyAndRunTest extends AbstractBuilderParticipantTest {
+
+	// the id of the test project
+	private static final String CLIENT = "Client";
+	// the module to be executed
+	private static final String MODULE_TO_RUN = "A";
+
+	@Inject
+	private Provider<RunnableInstallDependencies> installDependenciesRunnable;
+
+	@Inject
+	private ShippedCodeInitializeTestHelper shippedCodeInitializeTestHelper;
+
+	@Inject
+	private ExternalLibrariesActionsHelper externals;
+
+	@Inject
+	private StatusHelper statusHelper;
+
+	@Inject
+	private RuntimeEnvironmentsHelper runtimeEnvironmentsHelper;
+
+	@Inject
+	private IN4JSCore n4jsCore;
+
+	@Before
+	@Override
+	public void setUp() throws Exception {
+		super.setUp();
+		shippedCodeInitializeTestHelper.setupBuiltIns();
+	}
+
+	@After
+	@Override
+	public void tearDown() throws Exception {
+		// clear library manager to avoid confusing tests being executed after this test class
+		final MultiStatus multistatus = statusHelper
+				.createMultiStatus("Status of deleting NPM packages from library manager.");
+		externals.maintenanceDeleteNpms(multistatus);
+
+		shippedCodeInitializeTestHelper.tearDownBuiltIns();
+		super.tearDown();
+	}
+
+	/**
+	 * See JavaDoc of class.
+	 */
+	@Test
+	public void testRunWithCycleInDependencyTree() throws CoreException {
+		final IProject project = createJSProject(CLIENT, "src", "src-gen",
+				b -> b
+						// test project of type 'library'
+						.withType(ProjectType.LIBRARY)
+						.withDependency("mocha", "*")
+						.withDependency("eslint", "*"));
```

Comment:
Since the test relies on the internals of mocha and eslint (i.e. having a dependency cycle between them) it's important to use fixed versions of these projects. Otherwise the mocha/eslint devs might render our tests ineffective at any time by releasing a new version.